extensions/http-curl/client/HTTPCallback.h

```diff
@@ -33,148 +34,195 @@ namespace minifi {
   }
 
-  virtual int64_t process(std::shared_ptr<io::BaseStream> stream) {
-
+  int64_t process(std::shared_ptr<io::BaseStream> stream) override {
     std::vector<char> vec;
 
     if (stream->getSize() > 0) {
       vec.resize(stream->getSize());
-
       stream->readData(reinterpret_cast<uint8_t*>(vec.data()), stream->getSize());
     }
 
-    size_t added_size = vec.size();
-
-    byte_arrays_.enqueue(std::move(vec));
-
-    cv.notify_all();
-
-    return added_size;
-
+    return processInner(std::move(vec));
   }
 
-  virtual int64_t process(uint8_t *vector, size_t size) {
-
+  virtual int64_t process(const uint8_t* data, size_t size) {
     std::vector<char> vec;
+    vec.resize(size);
+    memcpy(vec.data(), reinterpret_cast<const char*>(data), size);
 
-    if (size > 0) {
-      vec.resize(size);
+    return processInner(std::move(vec));
+  }
 
-      memcpy(vec.data(), vector, size);
+  void write(std::string content) override {
+    std::vector<char> vec;
+    vec.assign(content.begin(), content.end());
 
-      size_t added_size = vec.size();
+    (void) processInner(std::move(vec));
+  }
 
-      byte_arrays_.enqueue(std::move(vec));
+  char* getBuffer(size_t pos) override {
+    logger_->log_trace("getBuffer(pos: %zu) called", pos);
 
-      cv.notify_all();
+    std::unique_lock<std::mutex> lock(mutex_);
 
-      return added_size;
-    } else {
-      return 0;
+    seekInner(lock, pos);
+    if (ptr == nullptr) {
+      return nullptr;
     }
 
-  }
+    size_t relative_pos = pos - current_buffer_start_;
+    current_pos_ = pos;
 
-  virtual void write(std::string content) {
-    std::vector<char> vec;
-    vec.assign(content.begin(), content.end());
-    byte_arrays_.enqueue(vec);
+    return ptr + relative_pos;
   }
 
-  virtual char *getBuffer(size_t pos) {
-
-    // if there is no space remaining in our current buffer,
-    // we should load the next. If none exists after that we have no more buffer
-    std::lock_guard<std::recursive_mutex> lock(mutex_);
+  const size_t getRemaining(size_t pos) override {
+    logger_->log_trace("getRemaining(pos: %zu) called", pos);
 
-    if ((pos - previous_pos_) >= current_vec_.size() || current_vec_.size() == 0)
-      load_buffer();
+    std::unique_lock<std::mutex> lock(mutex_);
+    seekInner(lock, pos);
+    return total_bytes_loaded_ - pos;
+  }
 
-    if (ptr == nullptr)
-      return nullptr;
+  const size_t getBufferSize() override {
+    logger_->log_trace("getBufferSize() called");
 
-    size_t absolute_position = pos - previous_pos_;
+    std::unique_lock<std::mutex> lock(mutex_);
+    // This is needed to make sure that the first buffer is loaded
+    seekInner(lock, current_pos_);
+    return total_bytes_loaded_;
+  }
 
-    current_pos_ = pos;
+ private:
 
-    return ptr + absolute_position;
+  /**
+   * Loads the next available buffer
+   * @param lock unique_lock which *must* own the lock
+   */
+  inline void loadNextBuffer(std::unique_lock<std::mutex>& lock) {
```

Comment:
I usually use `assert` to document assumptions. Feel free to ignore this if you prefer it that way.

Reply:
I prefer it this way (or by really acting on the assertion, like throwing an exception, but I didn't want to do this in this case).